net/mptcp/mptcp_blest.c

```diff
@@ -0,0 +1,476 @@
+	 * during the slow flows last RTT => increase lambda
+	 * otherwise decrease
+	 */
+	if (blest_cb->retrans_flag) {
+		/* need to slow down on the slow flow */
+		blest_cb->lambda_1000 += dyn_lambda_bad;
+	} else {
+		/* use the slow flow more */
+		blest_cb->lambda_1000 -= dyn_lambda_good;
+	}
+	blest_cb->retrans_flag = false;
+
+	/* cap lambda_1000 to its value range */
+	if (blest_cb->lambda_1000 > max_lambda * 100)
+		blest_cb->lambda_1000 = max_lambda * 100;
+	else if (blest_cb->lambda_1000 < min_lambda * 100)
+		blest_cb->lambda_1000 = min_lambda * 100;
+
+	blest_cb->last_lambda_update = tcp_jiffies32;
+}
+
+/* how many bytes will sk send during the rtt of another, slower flow? */
+static u32 blestsched_estimate_bytes(struct sock *sk, u32 time_8)
+{
+	struct tcp_sock *tp = tcp_sk(sk);
+	struct blestsched_priv *blest_p = blestsched_get_priv(tp);
+	struct blestsched_cb *blest_cb = blestsched_get_cb(mptcp_meta_tp(tp));
+	u32 avg_rtt, num_rtts, ca_cwnd, packets;
+
+	avg_rtt = (blest_p->min_srtt_us + blest_p->max_srtt_us) / 2;
+	if (avg_rtt == 0)
+		num_rtts = 1; /* sanity */
+	else
+		num_rtts = (time_8 / avg_rtt) + 1; /* round up */
+
+	/* during num_rtts, how many bytes will be sent on the flow?
+	 * assumes for simplification that Reno is applied as congestion-control
+	 */
+	if (tp->snd_ssthresh == TCP_INFINITE_SSTHRESH) {
+		/* we are in initial slow start */
+		if (num_rtts > 16)
+			num_rtts = 16; /* cap for sanity */
+		packets = tp->snd_cwnd * ((1 << num_rtts) - 1); /* cwnd + 2*cwnd + 4*cwnd */
+	} else {
+		ca_cwnd = max(tp->snd_cwnd, tp->snd_ssthresh + 1); /* assume we jump to CA already */
+		packets = (ca_cwnd + (num_rtts - 1) / 2) * num_rtts;
+	}
+
+	return div_u64(((u64)packets) * tp->mss_cache * blest_cb->lambda_1000, 1000);
+}
+
+static u32 blestsched_estimate_linger_time(struct sock *sk)
+{
+	struct tcp_sock *tp = tcp_sk(sk);
+	struct blestsched_priv *blest_p = blestsched_get_priv(tp);
+	u32 estimate, slope, inflight, cwnd;
+
+	inflight = tcp_packets_in_flight(tp) + 1; /* take into account the new one */
+	cwnd = tp->snd_cwnd;
+
+	if (inflight >= cwnd) {
+		estimate = blest_p->max_srtt_us;
+	} else {
+		slope = blest_p->max_srtt_us - blest_p->min_srtt_us;
+		if (cwnd == 0)
+			cwnd = 1; /* sanity */
+		estimate = blest_p->min_srtt_us + (slope * inflight) / cwnd;
+	}
+
+	return (tp->srtt_us > estimate) ? tp->srtt_us : estimate;
+}
+
+/* This is the BLEST scheduler. This function decides on which flow to send
+ * a given MSS. If all subflows are found to be busy or the currently best
+ * subflow is estimated to possibly cause HoL-blocking, NULL is returned.
+ */
+struct sock *blest_get_available_subflow(struct sock *meta_sk, struct sk_buff *skb,
+					 bool zero_wnd_test)
+{
+	struct mptcp_cb *mpcb = tcp_sk(meta_sk)->mpcb;
+	struct sock *bestsk, *minsk;
+	struct tcp_sock *meta_tp, *besttp;
+	struct mptcp_tcp_sock *mptcp;
+	struct blestsched_priv *blest_p;
+	u32 min_srtt = 0xffffffff;
```

Comment:
detail: there are macros for that: `U32_MAX` (same below)

Reply:
> detail: there are macros for that: `U32_MAX` (same below)

Thanks for the hint @matttbe ! I gonna use the macro instead. Please tell me if I should do a seperate commit once @cpaasch has already approved the PR or perform a force-push?

Thanks for your Review :)

Reply:
if it is not merged yet, feel free to override this commit  ;-)

Reply:
> feel free to override this commit ;-)

@matttbe I have overridden the commit now to use the `U32_MAX`. It now also includes a fix for the compiler warning that `minsk` could be uninitialized.

Reply:
good catch!
Thx for the update!

https://github.com/multipath-tcp/mptcp/compare/91a73ec36ea427307f5cf55d02f7474fd455fa2f..e34cb523a8ca561d1cef8d85c23f90295d08ef13